openvm/src/customize_exe.rs

```diff
@@ -571,83 +567,140 @@ fn sort_blocks_by_pgo_cell_cost_and_cache_apc<P: IntoOpenVm>(
             )
             .ok()?;
 
-            // calculate cells saved per row
-            let apc_cells_per_row = apc_cache_entry.autoprecompile.unique_references().count();
-            let original_cells_per_row: usize = acc_block
+            // compute cost and cells_saved_per_row
+            let apc_cells = entry.autoprecompile.unique_references().count();
+            let orig_cells: usize = block
                 .statements
                 .iter()
-                .map(|instruction| {
-                    air_width_by_opcode
-                        .get(&instruction.opcode.as_usize())
-                        .unwrap()
-                })
+                .map(|instr| air_width_by_opcode[&instr.opcode.as_usize()])
                 .sum();
-            let cells_saved_per_row = original_cells_per_row - apc_cells_per_row;
-            assert!(cells_saved_per_row > 0);
-            tracing::debug!(
-                "Basic block start_idx: {}, cells saved per row: {}",
-                acc_block.start_idx,
-                cells_saved_per_row
-            );
-
-            Some((
-                (acc_block.start_idx, cells_saved_per_row),
-                (acc_block.start_idx, apc_cache_entry),
-            ))
+            let cells_saved = orig_cells - apc_cells;
+            let freq = *pgo_program_idx_count
+                .get(&(block.start_idx as u32))
+                .unwrap_or(&0) as usize;
+            let cost = (cells_saved * freq) as u32;
+
+            // build a 1-element heap
+            let mut local = BinaryHeap::new();
+            local.push(Node {
+                cost,
+                start: block.start_idx,
+                entry,
+                cells_saved,
+            });
+            Some(local)
         })
+        .reduce(
+            // identity: empty heap
+            BinaryHeap::new,
+            // merge two heaps, pruning back to max_cache
+            |mut a, mut b| {
+                for node in b.drain() {
+                    a.push(node);
+                    if a.len() > max_cache {
+                        a.pop();
+                    }
+                }
+                a
+            },
+        );
+
+    // created from the min heap, so should be sorted already
+    let (cells_saved_by_bb, new_apc_cache): (
+        HashMap<usize, usize>,
+        HashMap<usize, CachedAutoPrecompile<P>>,
+    ) = top_heap
+        .into_par_iter()
+        .map(
+            |Node {
+                 start,
+                 entry,
+                 cells_saved,
+                 ..
+             }| { ((start, cells_saved), (start, entry)) },
+        )
         .unzip();
 
     apc_cache.extend(new_apc_cache);
 
-    // filter out the basic blocks where the apc generation errored out
-    blocks.retain(|b| cells_saved_per_row_by_bb.contains_key(&b.start_idx));
+    // filter to the top `max_cache` blocks by cost
+    blocks.retain(|b| cells_saved_by_bb.contains_key(&b.start_idx));
```

Comment:
Can't we avoid this by putting the block inside the `Node` as well? Then the Nodes are already ranked by cells saved, and we just have to partition them, putting the apc in the cache and the block in the output?